src/metrics/counter.rs

```diff
@@ -171,6 +175,62 @@ impl<N, A> TypedMetric for Counter<N, A> {
     const TYPE: MetricType = MetricType::Counter;
 }
 
+impl<N, A> EncodeMetric for Counter<N, A>
+where
+    N: EncodeValue,
+    A: Atomic<N>,
+{
+    fn encode(&self, encoder: MetricEncoder) -> Result<(), std::fmt::Error> {
+        // TODO: Would be better to use never type instead of `()`.
+        encode_counter_with_maybe_exemplar::<(), N, f64>(self.get(), None, encoder)
+    }
+
+    fn metric_type(&self) -> MetricType {
+        Self::TYPE
+    }
+}
+
+// TODO: Import is off on CounterWithExemplar, should the type be moved here?
+// TODO: S, V, N, A are hard to grasp.
+impl<S, N, A> EncodeMetric for crate::metrics::exemplar::CounterWithExemplar<S, N, A>
+where
+    S: EncodeLabelSet,
+    N: EncodeValue + EncodeExemplarValue + Clone,
+    A: Atomic<N>,
+{
+    fn encode(&self, encoder: MetricEncoder) -> Result<(), std::fmt::Error> {
+        let (value, exemplar) = self.get();
+        encode_counter_with_maybe_exemplar(value, exemplar.as_ref(), encoder)
+    }
+
+    fn metric_type(&self) -> MetricType {
+        Counter::<N, A>::TYPE
+    }
+}
+
+fn encode_counter_with_maybe_exemplar<S, CounterValue, ExemplarValue>(
+    value: CounterValue,
+    exemplar: Option<&crate::metrics::exemplar::Exemplar<S, ExemplarValue>>,
+    mut encoder: MetricEncoder,
+) -> Result<(), std::fmt::Error>
+where
+    S: EncodeLabelSet,
+    CounterValue: EncodeValue,
```

Comment:
Instead of dispatching to the value here, we could have one function on `MetricEncoder` per type and we'd probably just need to `macro_rules` the counter definition for each supported value type.

Not sure if worth it but I think it would reduce some indirection in dispatching between all these things.